File: C++/uasbinary.cpp
#include <iostream>
#include <string>
using namespace std;

struct Node {
    string data;
    Node* left;
    Node* right;

    Node(string item) {
        data = item;
        left = nullptr;
        right = nullptr;
    }
};

class SilsilahKeluarga {
    private:
        Node* root;

    public:
        SilsilahKeluarga() {
            root = nullptr;
        }

    void tambah(string induk, string anak, string hubungan) {
        if (root == nullptr) {
            root = new Node(induk);
            cout << "Induk Utama: " << induk << endl;
        }

        Node* nodeBr = new Node(anak);
        if (hubungan == "kiri") {
            Node* indukNode = cariNode(root, induk);
            if (indukNode != nullptr) {
                if (indukNode->left == nullptr) {
                    indukNode->left = nodeBr;
                    cout << anak << " adalah anak kiri dari " << induk << endl;
                } else {
                    cout << "Anak kiri dari " << induk << " sudah ada" << endl;
                }
            } else {
                cout << induk << " tidak ditemukan dalam silsilah keluarga" << endl;
            }
        } else if (hubungan == "kanan") {
            Node* indukNode = cariNode(root, induk);
            if (indukNode != nullptr) {
                if (indukNode->right == nullptr) {
                    indukNode->right = nodeBr;
                    cout << anak << " adalah anak kanan dari " << induk << endl;
                } else {
                    cout << "Anak kanan dari " << induk << " sudah ada" << endl;
                }
            } else {
                cout << induk << " tidak ditemukan dalam silsilah keluarga" << endl;
            }
        } else {
            cout << "Hubungan tidak valid" << endl;
        }
    }

        void tampilkanSilsilahKeluarga() {
            cout << "Silsilah Keluarga:" << endl;
            tampilkanNode(root, 0);
        }

    private:
        Node* cariNode(Node* nodeSkrg, string item) {
            if (nodeSkrg == nullptr) {
                return nullptr;
            }

            if (nodeSkrg->data == item) {
                return nodeSkrg;
            }

            Node* hasilKiri = cariNode(nodeSkrg->left, item);
            if (hasilKiri != nullptr) {
                return hasilKiri;
            }

            Node* hasilKanan = cariNode(nodeSkrg->right, item);
            if (hasilKanan != nullptr) {
                return hasilKanan;
            }

            return nullptr;
        }

    void tampilkanNode(Node* nodeSkrg, int level) {
        if (nodeSkrg == nullptr) {
            return;
        }

        tampilkanNode(nodeSkrg->right, level + 1);

        for (int i = 0; i < level; i++) {
            cout << "    ";
        }
        cout << nodeSkrg->data << endl;

        tampilkanNode(nodeSkrg->left, level + 1);
    }
};

int main() {
    SilsilahKeluarga silsilah;
    int pilihan;
    string induk, anak, hubungan;

    do {
        cout << "Menu Silsilah Keluarga:" << endl;
        cout << "1. Tambah anggota keluarga" << endl;
        cout << "2. Tampilkan silsilah keluarga" << endl;
        cout << "3. Keluar" << endl;
        cout << "Masukkan pilihan Anda: ";
        cin >> pilihan;

        switch (pilihan) {
            case 1:
                cout << "Masukkan nama induk: ";
                cin >> induk;
                cout << "Masukkan nama anak: ";
                cin >> anak;
                cout << "Masukkan hubungan (kiri/kanan): ";
                cin >> hubungan;
                silsilah.tambah(induk, anak, hubungan);
                break;
            case 2:
                silsilah.tampilkanSilsilahKeluarga();
                break;
            case 3:
                cout << "Terimakasih!!" << endl;
                break;
            default:
                cout << "Pilihan tidak Ada" << endl;
        }

        cout << endl;
    } while (pilihan != 3);

    return 0;
}

File: java/uasbinary.java
package uas.java;

// iki dicopy ngisor e pakage
import java.util.Scanner;

/**
 * 1. hapus kabeh seng public class seng pakage karo autor ditokne
 * 
 */

 // iki kabeh dicopy
class Node {
    String data;
    Node left, right;

    Node(String item) {
        data = item;
        left = null;
        right = null;
    }
}

class SilsilahKeluarga {
    private Node root;

    public SilsilahKeluarga() {
        root = null;
    }

    void tambah(String induk, String anak, String hubungan) {
        if (root == null) {
            root = new Node(induk);
            System.out.println("Induk Utama : " + induk);
        }

        Node nodeBaru = new Node(anak);
        if (hubungan.equals("kiri")) {
            Node indukNode = cariNode(root, induk);
            if (indukNode != null) {
                if (indukNode.left == null) {
                    indukNode.left = nodeBaru;
                    System.out.println(anak + " Adalah Anak Kiri Dari " + induk );
                } else {
                    System.out.println("Anak Kiri Dari " + induk + " Sudah Ada");
                }
            } else {
                System.out.println(induk + " Tidak Ditemukan Dalam Silsilah Keluarga");
            } 
        } else if (hubungan.equals("kanan")) {
            Node indukNode = cariNode(root, induk);
            if (indukNode != null) {
                if (indukNode.right == null) {
                    indukNode.right = nodeBaru;
                    System.out.println(anak + " Adalah Anak Kanan Dari " + induk );
                } else {
                    System.out.println("Anak Kanan Dari " + induk + " Sudah Ada");
                }
            } else {
                System.out.println(induk + " Tidak Ditemukan Dalam Silsilah Keluarga");
            } 
        } else {
            System.out.println("Hubungan Tidak Valid");
        }
    }

    public void tampilkanSilsilahKeluarga() {
        System.out.println("Silsilah Keluarga : ");
        tampilkanNode(root, 0);
    }

    private Node cariNode(Node nodeSkrg, String item) {
        if (nodeSkrg == null) {
            return null;
        }

        if (nodeSkrg.data.equals(item)) {
            return nodeSkrg;
        }

        Node hasilKiri = cariNode(nodeSkrg.left, item);
        if (hasilKiri != null) {
            return hasilKiri;
        }

        Node hasilKanan = cariNode(nodeSkrg.right, item);
        if (hasilKanan != null) {
            return hasilKanan;
        }

        return null;
    }


    public void tampilkanNode(Node nodeSkrg, int level){
        if (nodeSkrg == null) {
            return;
        }

        tampilkanNode(nodeSkrg.right, level + 1);

        for (int i = 0; i < level; i++) {
            System.out.print("    ");
                      
        }
        System.out.println(nodeSkrg.data);

        tampilkanNode(nodeSkrg.left, level+1);
    }
}
// uasbinary iki dihanti karo jeneng file mu contone BinaryTree kemau gede cilik e podo
public class uasbinary {
   public static void main(String[] args) {
    SilsilahKeluarga silsilah = new SilsilahKeluarga();
    int pilihan;
    String induk, anak, hubungan;

    Scanner scanner = new Scanner(System.in);

        do {
            System.out.println("Menu Silsilah Keluarga:");
            System.out.println("1. Tambah anggota keluarga");
            System.out.println("2. Tampilkan silsilah keluarga");
            System.out.println("3. Keluar");
            System.out.print("Masukkan pilihan Anda: ");
            pilihan = scanner.nextInt();

            switch (pilihan) {
                case 1:
                    System.out.print("Masukkan nama induk: ");
                    induk = scanner.next();
                    System.out.print("Masukkan nama anak: ");
                    anak = scanner.next();
                    System.out.print("Masukkan hubungan (kiri/kanan): ");
                    hubungan = scanner.next();
                    silsilah.tambah(induk, anak, hubungan);
                    break;
                case 2:
                    silsilah.tampilkanSilsilahKeluarga();
                    break;
                case 3:
                    System.out.println("Terimakasih!!");
                    break;
                default:
                    System.out.println("Pilihan tidak Ada");
            }

            System.out.println();
        } while (pilihan != 3);
    }
}
